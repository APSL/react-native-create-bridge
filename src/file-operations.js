const path = require("path");
const fs = require("mz/fs");

exports.pkg = require(path.join(process.cwd(), "package.json"));

function getFileNames(dirPath) {
  return fs.readdir(dirPath).catch(e => console.error("[getFileNames] ", e));
}

function readFile(file, readDirPath) {
  return fs
    .readFile(path.join(readDirPath, file), "utf-8")
    .catch(e => console.error("[readFile] ", e));
}

function parseFile(fileData, templateName, packageName = null, app = null) {
  return fileData
    .replace(/{{template}}/g, templateName)
    .replace(/{{packageName}}/g, packageName)
    .replace(/{{app}}/g, app);
}

function readAndWriteFiles(
  files,
  paths,
  templateName,
  packageName = null,
  app = null
) {
  const { readDirPath, writeDirPath } = paths;
  return Promise.all(
    files.map(file => {
      readFile(file, readDirPath).then(fileData => {
        const parsedFile = parseFile(fileData, templateName, packageName, app);
        const fileName = file.replace("Template", templateName);
        return fs.writeFile(path.join(writeDirPath, fileName), parsedFile);
      });
    })
  ).catch(e => console.error("[readAndWriteFiles] ", e));
}

module.exports = {
  getFileNames,
  readFile,
  parseFile,
  readAndWriteFiles
};
