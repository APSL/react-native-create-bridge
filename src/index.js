#! /usr/bin/env node

import inquirer from 'inquirer'
import path from 'path'
import fs from 'mz/fs'
import isValid from 'is-valid-path'
import mkdir from 'mkdirp-promise'

const templateNameRegex = /\w+/
const pkg = require(path.join(process.cwd(), 'package.json'))
const promptConfig = [
  {
    type: 'input',
    name: 'templateName',
    message: 'What is your bridge module called?',
    default: 'ExampleBridge',
    validate: (input) => templateNameRegex.test(input),
  },
  {
    type: 'checkbox',
    name: 'environment',
    message: 'What OS & languages would you like to support?',
    default: ['Android/Java', 'iOS/Swift', 'iOS/Objective-C'],
    choices: ['Android/Java', 'iOS/Swift', 'iOS/Objective-C']
  },
  {
    type: 'input',
    name: 'jsPath',
    message: 'What directory should we deliver your JS files to?',
    default: '.',
    validate: (input) => isValid(input),
  }
]

const environmentMap = {
  'Android/Java': createAndroidEnvironment,
  'iOS/Swift': createSwiftEnvironment,
  'iOS/Objective-C': createObjCEnvironment
}

export function isRNVersionNew ({ dependencies }) {
  return parseFloat(dependencies["react-native"]) >= 0.4
}

async function init () {
  try {
    const answers = await inquirer.prompt(promptConfig)

    const promises = answers.environment.map(async function(env) {
      return environmentMap[env](answers.templateName)
    })

    promises.push(createJSEnvironment(answers.templateName, answers.jsPath))

    await Promise.all(promises)

    console.log('Your bridge was successfully created! ðŸŽ‰')
  } catch (e) {
    console.log(e)
  }
}

export function parseFile (fileData, templateName, packageName = null, app = null) {
  const isRNNew = isRNVersionNew(pkg)

  const iOSHeader = isRNNew ? '<React/' : '"'
  const iOSCloser = isRNNew ? '>' : '"'

  return fileData.replace(/{{template}}/g, templateName)
    .replace(/{{packageName}}/g, packageName)
    .replace(/{{app}}/g, app)
    .replace(/{{iOSHeader}}/g, iOSHeader)
    .replace(/{{iOSCloser}}/g, iOSCloser)
}

export async function readFile (file, readDirPath) {
  const fileData = await fs.readFile(path.join(readDirPath, file), 'utf-8')
  return fileData
}

function readAndWriteFiles (files, paths, templateName, packageName = null, app = null) {
  const { readDirPath, writeDirPath } = paths
  return Promise.all(
    files.map(async function(file) {
      const fileData = await readFile(file, readDirPath)
      const parsedFile = parseFile(fileData, templateName, packageName, app)
      const fileName = file.replace('Template', templateName)
      return fs.writeFile(path.join(writeDirPath, fileName), parsedFile)
    })
  )
}

async function createAndroidEnvironment (templateName) {
  const appPath = path.join(
    process.cwd(),
    'android',
    'app',
    'src',
    'main',
    'java',
    'com',
    pkg.name
  )
  const writeDirPath = await mkdir(path.join(appPath, templateName.toLowerCase()))
  const paths = {
    readDirPath: path.join(__dirname, '..', 'templates', 'android'),
    writeDirPath
  }
  const files = ['TemplatePackage.java', 'TemplateModule.java', 'TemplateManager.java']
  return readAndWriteFiles(files, paths, templateName, templateName.toLowerCase(), pkg.name)
}

function createSwiftEnvironment (templateName) {
  const paths = {
    readDirPath: path.join(__dirname, '..', 'templates', 'ios-swift'),
    writeDirPath: path.join(process.cwd(), 'ios')
  }
  const files = ['Template.m', 'TemplateManager.swift']
  return readAndWriteFiles(files, paths, templateName)
}

function createObjCEnvironment (templateName) {
  const paths = {
    readDirPath: path.join(__dirname, '..', 'templates', 'ios-objc'),
    writeDirPath: path.join(process.cwd(), 'ios')
  }
  const files = ['Template.h', 'TemplateManager.m']
  return readAndWriteFiles(files, paths, templateName)
}

async function createJSEnvironment (templateName, jsPath) {
  await mkdir(jsPath)

  const paths = {
    readDirPath: path.join(__dirname, '..', 'templates', 'js'),
    writeDirPath: jsPath
  }
  const files = ['TemplateNativeModule.js', 'TemplateNativeView.js']
  return readAndWriteFiles(files, paths, templateName)
}

init()
