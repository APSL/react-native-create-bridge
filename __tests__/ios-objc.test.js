import path from 'path'
import { readFile, parseFile } from '../src/file-operations'

describe('iOS/Obj-C', () => {
  const templateName = 'TestModule'
  const readDirPath = path.join(__dirname, '..', 'templates', 'ios-objc')
  const pkg = {
    dependencies: {
      'react-native': '0.4'
    }
  }

  it('creates a Template.h', async () => {
    const fileData = await readFile('Template.h', readDirPath)
    const parsedFile = parseFile(fileData, templateName)
    expect(parsedFile).toMatchSnapshot()
  })

  it('creates a TemplateManager.m', async () => {
    const fileData = await readFile('TemplateManager.m', readDirPath)
    const parsedFile = parseFile(fileData, templateName)
    expect(parsedFile).toMatchSnapshot()
  })
})
