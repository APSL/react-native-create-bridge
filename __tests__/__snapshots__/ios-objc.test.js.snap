exports[`iOS/Obj-C creates a Template.h 1`] = `
"//  Created by react-native-create-bridge

// import RCTBridgeModule
#if __has_include(<React/RCTBridgeModule.h>)
#import <React/RCTBridgeModule.h>
#elif __has_include(“RCTBridgeModule.h”)
#import “RCTBridgeModule.h”
#else
#import “React/RCTBridgeModule.h” // Required when used as a Pod in a Swift project
#endif

@interface TestModule : NSObject <RCTBridgeModule>
  // Define class properties here with @property
@end
"
`;

exports[`iOS/Obj-C creates a TemplateManager.m 1`] = `
"//  Created by react-native-create-bridge

#import \"TestModule.h\"

// import RCTBridge
#if __has_include(<React/RCTBridge.h>)
#import <React/RCTBridge.h>
#elif __has_include(“RCTBridge.h”)
#import “RCTBridge.h”
#else
#import “React/RCTBridge.h” // Required when used as a Pod in a Swift project
#endif

// import RCTEventDispatcher
#if __has_include(<React/RCTEventDispatcher.h>)
#import <React/RCTEventDispatcher.h>
#elif __has_include(“RCTEventDispatcher.h”)
#import “RCTEventDispatcher.h”
#else
#import “React/RCTEventDispatcher.h” // Required when used as a Pod in a Swift project
#endif

@implementation TestModule
@synthesize bridge = _bridge;

// Export a native module
// https://facebook.github.io/react-native/docs/native-modules-ios.html
RCT_EXPORT_MODULE();

- (NSDictionary *)constantsToExport
{
  return @{
           @\"EXAMPLE\": @\"example\"
         };
}

// Return the native view that represents your React component
- (UIView *)view
{
  return [[UIView alloc] init];
}

// Export methods to a native module
// https://facebook.github.io/react-native/docs/native-modules-ios.html
RCT_EXPORT_METHOD(exampleMethod)
{
  [self.emitMessageToRN:@\"EXAMPLE_EVENT\" :nil]
}

#pragma mark - Private methods

// Implement methods that you want to export to the native module
- (void) emitMessageToRN: (NSString *)eventName :(NSDictionary *)params {
  // The bridge eventDispatcher is used to send events from native to JS env
  // No documentation yet on DeviceEventEmitter: https://github.com/facebook/react-native/issues/2819
  [self.bridge.eventDispatcher sendAppEventWithName: eventName body: params];
}

@end
"
`;
